pkg/sharky/slots.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package sharky
+
+import (
+	"io"
+	"sync"
+)
+
+type slots struct {
+	data []byte          // byteslice serving as bitvector: i-t bit set <>
+	size uint32          // number of slots
+	head uint32          // the first free slot
+	file sharkyFile      // file to persist free slots across sessions
+	in   chan uint32     // incoming channel for free slots,
+	out  chan uint32     // outgoing channel for free slots
+	wg   *sync.WaitGroup // count started write operations
+}
+
+func newSlots(file sharkyFile, wg *sync.WaitGroup) *slots {
+	return &slots{
+		file: file,
+		in:   make(chan uint32),
+		out:  make(chan uint32),
+		wg:   wg,
+	}
+}
+
+// load inits the slots from file, called after init
+func (sl *slots) load() (err error) {
+	sl.data, err = io.ReadAll(sl.file)
+	if err != nil {
+		return err
+	}
+	sl.size = uint32(len(sl.data) * 8)
+	sl.head = sl.next(0)
+	return err
+}
+
+// save persists the free slot bitvector on disk (without closing)
+func (sl *slots) save() error {
+	if err := sl.file.Truncate(0); err != nil {
+		return err
+	}
+	if _, err := sl.file.Seek(0, 0); err != nil {
+		return err
+	}
+	if _, err := sl.file.Write(sl.data); err != nil {
+		return err
+	}
+	return sl.file.Sync()
+}
+
+// extend adapts the slots to an extended size shard
+// extensions are bytewise: can only be multiples of 8 bits
+func (sl *slots) extend(n int) {
```

Comment:
since `n` is always `1` we can get rid of the confusing argument

Reply:
We use this argument in Recovery mode. Keeping this for now